apps/lend/src/entities/token/lib/query-data.ts

```diff
@@ -0,0 +1,6 @@
+import { tokenKeys } from '../model'
+import { queryClient } from '@/shared/api/query-client'
+import { TokenKey } from '@/entities/token'
+
+export const getTokenQueryData = <Result, Key extends TokenKey = TokenKey>(key: TokenKey, keyData: Parameters<typeof tokenKeys[Key]>[0]) =>
```

Comment:
Take the generic factory function out to the shared modules to only specify the keys object here and have a special function for tokens.
```ts
export const getTokenQueryData = getQueryDataFactory(tokensKeys)
```

Reply:
Another architectural problem to think about is how to get the type of the return value depending on the key. To avoid using the `Result` type. This can be done on pure TS types

Reply:
yeah the result type is only defined by the query function, and then connected via query options. So it's not possible to get the return type just from the keys

Reply:
[That](https://github.com/curvefi/curve-frontend/pull/376/commits/4b69aae9eff2531738c5206dd0b8608064777b69) kind of makes things more complicated :disappointed:

Reply:
Is it? It's a once-written factory function that was written and forgotten. Or what are you talking about?